codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -464,4 +444,125 @@ public Http2Settings pollSentSettings() {
+     * only writes padding once for the entire payload as opposed to writing it once per-frame. This makes the
+     * {@link #size} calculation deterministic thereby greatly simplifying the implementation.
+     * <p>
+     * If frame-splitting is required to fit within max-frame-size and flow-control constraints we ensure that
+     * the passed promise is not completed until last frame write.
+     * </p>
+     */
+    private final class FlowControlledData extends FlowControlledPayloadBase {
+        private final ByteBuf data;
+        private int size;
+
+        private FlowControlledData(ChannelHandlerContext ctx, int streamId, ByteBuf data, int padding,
+                                    boolean endOfStream, ChannelPromise promise) {
+            super(ctx, padding, endOfStream, streamId, promise);
+            this.data = data;
+            size = data.readableBytes() + padding;
+        }
+
+        @Override
+        public int size() {
+            return size;
+        }
+
+        @Override
+        public void error(Throwable cause) {
+            data.release();
+            promise.tryFailure(cause);
+        }
+
+        @Override
+        public boolean write(int allowedBytes) {
+            if (allowedBytes == 0 && size() != 0) {
+                // No point writing an empty DATA frame, wait for a bigger allowance.
+                return false;
+            }
+            int maxFrameSize = frameWriter().configuration().frameSizePolicy().maxFrameSize();
+            try {
+                int bytesWritten = 0;
+                do {
+                    int allowedFrameSize = Math.min(maxFrameSize, allowedBytes - bytesWritten);
+                    ByteBuf toWrite;
+                    // Let data consume the frame before padding.
+                    if (data.readableBytes() == 0) {
+                        toWrite = Unpooled.EMPTY_BUFFER;
+                    } else  if (data.readableBytes() > allowedFrameSize) {
+                        toWrite = data.readSlice(Math.min(data.readableBytes(), allowedFrameSize)).retain();
+                    } else {
+                        toWrite = data;
+                    }
+                    int writeablePadding = Math.min(allowedFrameSize - toWrite.readableBytes(), padding);
+                    padding -= writeablePadding;
+                    bytesWritten += toWrite.readableBytes() + writeablePadding;
+                    frameWriter().writeData(ctx, streamId, toWrite, writeablePadding,
+                            size == bytesWritten && endOfStream,
+                            size == bytesWritten ? promise : ctx.voidPromise());
+                } while (size != bytesWritten && allowedBytes > bytesWritten);
+                size -= bytesWritten;
+                return true;
+            } catch (Throwable e) {
+                promise.tryFailure(e);
+                ReferenceCountUtil.safeRelease(data);
```

Comment:
I think safeRelease is the best course of action. The external expectation for normal operation is that the buffer is fully read and released on completion, no reason to vary from that for failure.

I'll add protection for write-after-error.